Refactor Hearth Dockerfile

trevorgowing opened the pull request:
* use node:14.17.5-apline3.14 as base image
* since libxmljs no longer produces releases built for the latest
  versions of node the package has to be built just in time
* move libxmljs to a productions dependency since it is a transitive
  dependency of fhir

trevorgowing committed:
Refactor Hearth Dockerfile

* use node:14.17.5-apline3.14 as base image
* since libxmljs no longer produces releases built for the latest
  versions of node the package has to be built just in time
* move libxmljs to a productions dependency since it is a transitive
  dependency of fhir

trevorgowing commented:
I thought that it would be better to rather define `libxmljs` as prod dependency rather than installing dev dependencies in the image.

I guess we could take it a step further and only copy the `libxmljs` package rather than copying all of the installed dependencies from the build image.

This seemed like the best middle ground.
Let me know your thoughts.

bausmeier commented on Dockerfile:
```diff
@@ -1,11 +1,15 @@
-FROM node:carbon
+FROM node:14.17.5 as build
```
Instead of using glibc like Debian does, Alpine uses musl libc which is only partly compatible, so it's a bad idea to compile native dependencies using one C standard library and run them on another. We should be building and running the app using the same Linux distribution.

trevorgowing replied:
How do you feel about this?

```
FROM node:14.17.5-alpine3.14 as build
RUN apk update && apk add --no-cache build-base python3 py3-pip
WORKDIR /app
COPY package.json yarn.lock /app/
RUN yarn install --frozen-lockfile --production && cd node_modules/libxmljs && yarn install

FROM node:14.17.5-alpine3.14
WORKDIR /app
COPY --from=build /app/node_modules /app/node_modules/
COPY package.json yarn.lock /app/
COPY config /app/config/
COPY lib /app/lib/
COPY scripts /app/scripts/
USER node
EXPOSE 3447
CMD ["node", "lib/server.js"]
```

bausmeier replied:
That looks good to me.

trevorgowing replied:
I am getting these warnings at build time both on debian and alpine.

Any of them concerning?

```
warning node-pre-gyp@0.11.0: Please upgrade to @mapbox/node-pre-gyp: the non-scoped node-pre-gyp package is deprecated and only the @mapbox scoped package will recieve updates in the future
warning nodeunit@0.11.3: you are strongly encouraged to use other testing options
warning nodeunit > tap > coveralls > request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
warning nodeunit > tap > nyc > uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
warning nodeunit > tap > coveralls > request > uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
warning nodeunit > tap > coveralls > request > har-validator@5.1.5: this library is no longer supported
```

bausmeier replied:
No. Other than `node-pre-gyp` those are only for the `nodeunit` package which is used for testing, and isn't included in the final image. The `node-pre-gyp` warning is valid, but I think we'll have to upgrade a bunch of other packages if we want to resolve that.

bausmeier commented on package.json:
```diff
@@ -31,6 +31,7 @@
     "express-xml-bodyparser": "^0.3.0",
     "fhir": "^3.3.1",
     "jsonwebtoken": "^8.2.2",
+    "libxmljs": "^0.19.7",
```
This shouldn't be a production dependency. It's only used directly in the tests, so it should be a dev dependency. If the `fhir` package were to stop depending on `libxmljs` (like it has in the newer versions), then we'd end up with an extra package in our dependencies which is not actually used. We should rather let Yarn decide whether or not it needs to be installed based on the dependencies of our other production dependencies.

The issue which you were seeing seems to be a bug in Yarn when using a combination of resolutions for production dependencies and the --production flag, causing the `libxmljs` package not to be built. Adding ` && cd node_modules/libxmljs && yarn install` on the end of the  `yarn install --frozen-lockfile --production` command to manually run the build for that package solves it.

trevorgowing replied:
How did you manage to figure that out?

bausmeier replied:
By shelling into the Docker container after it had been built, and exploring the node_modules directory. I also switched from yarn to npm locally to confirm that it's a yarn issue. That would be an alternative solution if we wanted to consider that.

trevorgowing replied:
And you saw that libxmljs was missing?

bausmeier replied:
No, I could see from the log output (or lack thereof) during install, and the error message on startup that it wasn't being built. I shelled into the container to figure out how to trigger the build for the package which was already installed to node_modules.

bausmeier replied:
It already gets hoisted from the fhir package into `node_modules/libxmljs`. The bug is just that the build step doesn't get run for that package when using the `--production` flag to `yarn install`.

trevorgowing committed:
Move libxmlsjs back to dev depdendencies

* libxmljs is only used by the tests even though it is also a transitive
  dependency of fhir

trevorgowing committed:
Change build base image to alpine

* alpine and debian linux use different c libraries which are not
  completely compatible so it is not recommended to build and run
  libraries on different linux distributions

trevorgowing committed:
Explictly install libxmljs

* There is a bug in yarn which is causing libxmljs not to be built when
  using the --production flag

trevorgowing committed:
Github Actions Docker Image Build Pipeline
